Feature: CustomScoreboard + Color Month in Vanilla/Custom Scoreboard

j10a1n15 opened the pull request:
# **GOT REPLACED WITH #893**







![image](https://github.com/hannibal002/SkyHanni/assets/45315647/25effa9c-3a9c-4dd1-aaf1-f96943f17f33)

Warning: Images are outdated again, will probably not update it

## Noteable Additions:
- Custom Scoreboard
- Colored months in Scoreboard (vanilla & custom)
- fixed ScoreboardData.sidebarlinesFormatted randomly cutting of chars
- /shcopyactionbar
- ~~shader to render a rounded rectangle~~ implemented with #851
- Maxwell api
  - All Power Stones 
  - Currently selected Power
- QuiverAPI
  - All Arrow Types
  - Currently selected Arrow
  - Currently Selected Arrow Amount
  - amount of all arrow types
- BitsAPI
  - all fame ranks with important informations
  - current amount of bits
  - bits left to claim
  - the players fame rank

I love apis

~~This pr requires [#83](https://github.com/hannibal002/SkyHanni-REPO/pull/83) to be merged~~ got merged ty!


# What
A custom skyblock scoreboard with multiple options

# Options
> ## Main Toggle
> - What: Toggles the scoreboard
> - Default: Off

> ## Dragable list of
> > ![image](https://github.com/hannibal002/SkyHanni/assets/45315647/cae0654e-e930-41b6-9ba3-9121b0525eec)
> > ![image](https://github.com/hannibal002/SkyHanni/assets/45315647/3b2e3d5e-f1dd-4779-9388-f6b3f6197fcf)
> > 
> > Note: Hidden lines are only hidden when the option "Hide Info not relevant to location" is enabled; Empty lines are only hidden when "Hide lines with no info" is enabled; The Style depends on if the option "Display Numbers First" is enabled.
> > ### Title
> > - What: A Title with a text input field to customise the text in the line
> > - Style: -
> > - Hidden: never
> > ### Profile
> > - What: Contains your profile icon (none if normal profile) and your profile name
> > - Style: Always icon then name
> > - Hidden: never
> > ### Purse
> > - What: Contains your current purse 
> > - Style: "Purse: num" & "num Purse
> > - Hidden: While in the Rift; No Purse
> > ### Motes
> > - What: Contains your current motes
> > - Style: "Motes: num" & "num Motes"
> > - Hidden: Hidden everywhere except rift; No Motes
> > ### Bank
> > - What: Contains your current bank, always rounded to the highest number type 
> > - Style: "Bank: num" & "num Bank
> > - Hidden: While in the Rift; No Bank
> > ### Bits
> > - What: Contains your current Bits
> > - Style: "Bits: num" & "num Bits"
> > - Hidden: While in the Rift; No Bits
> > ### Copper
> > - What: Contains your current Copper
> > - Style: "Copper: num" & "num Copper"
> > - Hidden: everywhere except Garden; No Copper
> > ### Gems
> > - What: Contains your current Gems
> > - Style: "Gems: num" & "num Gems"
> > - Hidden: While in the Rift; No Gems
> > ### Heat:
> > - What: Contains your current Heat
> > - Style: "Heat: <symbol> num" & "<symbol> num Heat"
> > - Hidden: Everywhere except Crystal Hollows; When empty
> > ### Empty line
> > ### Location
> > - What: Displays your current Location
> > - Style: Like Skyblocks Line
> > - Hidden: Never
> > ### Lobby Code
> > - What: Displays the current Lobby Code
> > - Style: its just the code idk
> > - Hidden: never
> > ### Ingame Date
> > - What: Displays the current ingame date with year
> > - Style: Late Summer 11th, Year 311
> > - Hidden: never
> > ### Current Event
> > Look below
> > ### Empty Line
> > ### Objective
> > - What: Displays the objective/quest like hypixel does
> > - Style: Like Hypixel
> > - Hidden: when no quest
> > ### Slayer
> > - What: Displays the current ongoing slayer
> > - Style: "Slayer\n - Slayer Tier\n - Kill Progress
> > - Hidden: everywhere except Rift, Park, Hub, End, Crimson; When empty
> > ### Powder
> > - What: Displays your current Powder
> > - Style: "Powder\n - Mithril: num\n - Gemstone: num"; Numbers and Powder type can also be switched
> > - Hidden: Everywhere except dwarven & crystal hollows
> > ### Mayor
> > - What: Displays the current mayor (& perks)
> > - Style: Depending on setting: Either only the current mayor or the current mayor + their perks
> > - Hidden: Only in the Rift
> > ### Empty Line
> > ### Party
> > - What: Displays the current party
> > - Style: "Party (count)" and then a new line for every nember till the setting "Max Party List" is hit
> > - Hidden: Everywhere except Dhub, Dungeons, Crimson, Kuudra (debatable); When empty
> > ### Maxwell Power
> > - What: The current selected Maxwell Power
> > - Style: Either "Power: <power>" & "<power> Power"
> > - Hidden: Only in the Rift
> > ### Footer
> > - What: A Footer with a text input field to customise the text in the line
> > - Style: -
> > - Hidden: never


# Events
> ### NONE
> - Shows nothing or "No Event"
> ### SERVER_CLOSE
> - shows the server closing
> ### DUNGEONS
> - Shows key, time elapsed, cleared, teammates
> ### KUUDRA
> - Shows the normal things you see in a kuudra run (i dont run kuudra so it could be broken)
> ### JACOB_CONTEST
> - Shows the current contest, like hypixel does
> ### JACOB_MEDALS
> - Shows the medals you see when opening anita
> ### WINTER
> - Shows the ongoing vulcano cube event thing
> ### MARINA
> - Shows the ongoing fishing festival
> ### NEW_YEAR
> - Shows the ongoing new year
> ### ORINGO
> - Shows the ongoing oringo pet guy thing
> ### MINING_EVENTS
> - Shows ongoing mining fiesta, wind compass, better together & mithril
> ### DAMAGE
> - Shows the current dmg to a dragon & protector
> ### ESSENCE
> - Shows the essence from the sb


> # Display Options
> > ### Hide Vanilla Scoreboard
> > What: Hides the vanilla scoreboard
> > Default: False
> > ### Display Numbers First
> > What: Displays numbers before the text
> > Default: False
> > ### Align to the right
> > What: Align scoreboard to the right
> > Default: false
> > ### Align vertical center
> > What: Align scoreboard to the vertical center
> > Default: false
> > ### Custom Title & Custom Footer
> > What: 2 input fields of respectively title and footer
> > Default: The default display stuff
> > ### Use Hypixels Title Animation
> > What: Use Hypixels Title Animation instead of the normal title
> > Default: False
> # Information Filtering
> > ### Hide Lines with no Info
> > What: Hides lines that contain no information
> > Default: True
> > ### Hide Info not relevant to location
> > What: Hides lines that are not relevant to location
> > Default: True
> > ### Hide consecutive empty lines
> > What: Hides consecutive empty lines in the scoreboard
> > Default: True
> # Background Options
> > ### Show scoreboard background
> > What: Background behind scoreboard
> > Default: true (i think)
> > ### Background Color
> > What: Customizeable background color
> > Default: smth similar to the vanilla background
> # Party Options
> > ### Max Party List
> > What: Sets a limit on how many party members are shown
> > Default: 4 of max 25
> > ### Show Party everywhere
> > What: Shows the party list everywhere
> > Default: false

> ### Show Mayor Perks
> What: Show the Perks of the current Mayor
> Default: True


# Backend changes
### Added /shcopyactionbar
### Added MaxwellAPI
### Maybe some more, I dont remember


# More Screenshots
Most amount of lines possible (please use hide irrelevant & empty lines):
![image](https://github.com/hannibal002/SkyHanni/assets/45315647/a3e50748-83cc-4466-8290-de9ef4040f45)

Most amount of lines, display numbers first false
![image](https://github.com/hannibal002/SkyHanni/assets/45315647/24a7d5e9-cab6-4040-9bfb-5e8a69ccc0ec)

Rift Moment
![image](https://github.com/hannibal002/SkyHanni/assets/45315647/cd0461d1-94df-4a25-b9fa-1c181a7c6731)

Custom Background Image
![image](https://github.com/hannibal002/SkyHanni/assets/45315647/5cf9e515-e480-42bc-90d4-0a12534114ad)


# Discord Suggestion
[By alpaka8123](https://discord.com/channels/997079228510117908/1162844830360146080)


### ty cimbraien for the pr style idea

j10a1n15 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/Events.kt:
```diff
@@ -0,0 +1,170 @@
+
+            // Mining Fiesta
+            if (TabListData.getTabList().any { it.startsWith("§6Mining Festival§f") }) {
+                list += "§6Mining Fiesta: " + TabListData.getTabList().nextAfter("§e§lEvent: §r§6Mining Fiesta")?.removePrefix(" Ends In: ")
```
Prob also gets shown when it starts soon

j10a1n15 commented:
Idk what your plans are, but I think we can go into public testing soon

walkerselby commented on src/main/java/at/hannibal2/skyhanni/test/command/CopyActionBar.kt:
```diff
@@ -0,0 +1,18 @@
+object CopyActionBar {
+    fun command(args: Array<String>) {
+        val noColor = args.size == 1 && args[0] == "true"
+        var string = ""
+        string = ActionBarStatsData.actionBar.transformIf({noColor}) { removeColor() }
```
Can merge these lines

j10a1n15 replied:
Huh yeah, why is it like that

walkerselby commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/ScoreboardElements.kt:
```diff
@@ -0,0 +1,470 @@
+    private val configLine: String
+) {
+    TITLE(
+        {
+            val alignment = if (config.displayConfig.centerTitleAndFooter) {
+                AlignmentEnum.CENTER
+            } else {
+                AlignmentEnum.LEFT
+            }
+
+            when (config.displayConfig.useHypixelTitleAnimation) {
+                true -> listOf(ScoreboardData.objectiveTitle to alignment)
+                false -> listOf(config.displayConfig.customTitle.get().toString().replace("&", "§") to alignment)
+            }
+        },
+        {
+            true
+        },
```
For the Scoreboard elements, could you extract these to functions to make it more readable?

j10a1n15 committed:
fixed pattern names

j10a1n15 committed:
fixed time to next mayor when date is after election in the same year

j10a1n15 committed:
removed old import

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
Added option to color months

j10a1n15 committed:
Fixed maxwellapi with new power

j10a1n15 committed:
Changed season colors to the colors of the atmospheric filter

j10a1n15 committed:
Added dark auction

j10a1n15 committed:
Fixed unknown lines causing fps drops

j10a1n15 committed:
Added effigies

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
small changes

j10a1n15 committed:
Redid whole config layout

j10a1n15 committed:
Fixed weird scoreboard lines when server closing

j10a1n15 committed:
Fixed magma boss detection

j10a1n15 committed:
The boss is reforming!

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
Multiline objective in kuudra

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
renamed extra -> unknown

j10a1n15 committed:
Refactor Events enum in Events.kt and InformationGetter class

This commit refactors the Events enum in Events.kt by removing the MAGMA_BOSS case and adding a new case for MAGMA_BOSS with additional functionality. It also updates the InformationGetter class to handle the new changes in the Events enum.

j10a1n15 committed:
Refactor getEventsDisplayPair function to handle empty Events list

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
Added garden pasting

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
Hypixel is smoking

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
Update EXTRA display message

j10a1n15 committed:
Added barn pasting

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
fixed that weird redstone line

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
added future todo

j10a1n15 committed:
Added quiver display

j10a1n15 committed:
Switched to repo pattern

j10a1n15 committed:
Removed debug chat msgs

j10a1n15 committed:
Now using RepoPattern.group()

j10a1n15 committed:
Updated todo

j10a1n15 committed:
Fixed var name, made it const

j10a1n15 committed:
Added BitsAPI

j10a1n15 committed:
Added option to see unclaimed bits

j10a1n15 committed:
Changed some wording, moved location

j10a1n15 committed:
Better fallback msg when no arrowtype is known

j10a1n15 committed:
Fixed error in savearrow()

j10a1n15 committed:
Idk those are "arrows" aswell

j10a1n15 committed:
Added raffle to mining event

j10a1n15 committed:
Fixed some things inside the mining events thing

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
fixed mithril gourmand event

j10a1n15 committed:
removed a useless thing

j10a1n15 committed:
Added current fame rank to bitsAPI

j10a1n15 committed:
Formatting fixes

j10a1n15 committed:
fixed with the scoreboard color fix thing

j10a1n15 committed:
moved scoreboard category to the top

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
Added separators to arrow display

j10a1n15 committed:
Hide quiver while on rift

j10a1n15 committed:
Hide quiver while on rift

j10a1n15 committed:
extract lobby code is now using repopattern!

j10a1n15 committed:
added goblin raid (im not testing that screw you)

j10a1n15 committed:
removed duplicate char from wind

j10a1n15 committed:
fixed repo pattern being not loaded correctly

j10a1n15 committed:
adde empty line

j10a1n15 committed:
Bits now show in rift

j10a1n15 committed:
fixed arrow remove thing

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
now loading inf quiv multiplier from repo

j10a1n15 committed:
changed bits to use bitapi

hannibal002 committed:
Merge branch 'beta' into customscoreboard

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/AlignmentConfig.java:
```diff
@@ -0,0 +1,20 @@
+    @Expose
+    @ConfigOption(name = "Align to the right", desc = "Align the scoreboard to the right side of the screen.")
+    @ConfigEditorBoolean
+    @FeatureToggle
```
Remove this and the other 18 feature toggle annotations. feature toggle is ment for standalone features. none of those settins will work when the scoreboard is disabled. Only one call should be kept for the main toggle.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/CustomScoreboardConfig.java:
```diff
@@ -0,0 +1,67 @@
+    )
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean enabled = false;
```
This is the main scoreabord toggle, here it is okey to have the feature toggle annotation set

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboard.kt:
```diff
@@ -0,0 +1,114 @@
+        }
+    }
+
+    private fun isCustomScoreboardEnabled() = config.enabled && LorenzUtils.inSkyBlock
```
Please always first check if the player is in sb, then check the config state. there is a small chance a race condition will cause an crash at this point when the check happens before the config is loaded.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboard.kt:
```diff
@@ -0,0 +1,114 @@
+    @SubscribeEvent
+    fun onChatMessage(event: LorenzChatEvent) {
+        // Reset Bits - We need this bc if another profile has 0 bits, it won't show the bits line
+        if (event.message.startsWith("§aYour profile was changed to:")) {
```
Maybe you can use ProfileJoinEvent here?

j10a1n15 replied:
Oh yeah I can switch back, it was broken at the time of implementation thats why I used this

hannibal002 replied:
I dont remember any bugs regarding profile join event getting fixed in the last months.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboard.kt:
```diff
@@ -0,0 +1,114 @@
+        InformationGetter.getInformation()
+    }
+
+    private fun createLines() = buildList<Pair<String, AlignmentEnum>> {
```
You can use `typealias SomeBetterName = Pair<String, AlignmentEnum>` in that file

j10a1n15 replied:
wtf thats so cool

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/ScoreboardElements.kt:
```diff
@@ -0,0 +1,413 @@
+    }
+
+    fun isVisible(): Boolean {
+        if (!config.informationFilteringConfig.hideIrrelevantLines) return true
```
Here you can wrap config.informationFilteringConfig into a new local config variable. if you dont use config else in this file, you can even replace it to config directly

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/CustomScoreboardConfig.java:
```diff
@@ -0,0 +1,67 @@
+    @ConfigOption(name = "Unknown Lines warning", desc = "Gives a chat warning when unknown lines are found in the scoreboard.")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean unknownLinesWarning = true;
```
this should maybe be disabled for non beta cases? or moved to dev config later? not sure

j10a1n15 replied:
idk you decide

j10a1n15 replied:
i mean its kinda like the normal bugs click to copy things, users are supposed to share them on the discord

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/InformationFilteringConfig.java:
```diff
@@ -0,0 +1,26 @@
+    public boolean hideEmptyLines = true;
+
+    @Expose
+    @ConfigOption(name = "Hide Info not relevant to location", desc = "Hide lines that are not relevant to the current location, like hiding copper while not in garden.")
```
This config name shold get too long/too small when looking at it

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/gui/GUIConfig.java:
```diff
@@ -23,6 +25,10 @@ public class GUIConfig {
     @ConfigEditorKeybind(defaultKey = Keyboard.KEY_NONE)
     public int keyBindOpen = Keyboard.KEY_NONE;
 
+    @Expose
```
Can we move this below global gui scale? Global gui scale feels more like a global setting to me.

j10a1n15 replied:
oh mb, it seems I cant read

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java:
```diff
@@ -120,6 +120,12 @@ public class MiscConfig {
+    @ConfigOption(name = "Color Month Names", desc = "Color the month names in the Scoreboard.\nAlso applies to the Custom Scoreboard")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean colorMonthNames = false;
```
This is a good use case of feature toggle annotation

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/Storage.java:
```diff
@@ -117,6 +120,34 @@ public static class ProfileSpecific {
         public String currentPet = "";
 
+        @Expose
+        public Powers currentPower = null;
```
The name is confusing at first look for me. can we give it a more descritive name? both the file name and the field name here

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/QuiverAPI.kt:
```diff
@@ -0,0 +1,200 @@
+
+private var infinityQuiverLevelMultiplier = 0.03f
+
+enum class Arrows(val arrow: String, val internalName: NEUInternalName) {
```
Please rename to ArrowType or something more enum-like

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt:
```diff
@@ -0,0 +1,103 @@
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+
+enum class Powers(val power: String) {
```
What power? magical power? also use enum name convention here

hannibal002 replied:
On a second thought, we might just move that into json and dont use a enum at all, should be more maintainable this way

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt:
```diff
@@ -0,0 +1,103 @@
+    ;
+}
+
+object MaxwellAPI {
```
Im not a big fan of multiple big classes in the same file atm. i dont have logical arguments against rn, it just doesnt match the current code base.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt:
```diff
@@ -0,0 +1,103 @@
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        val message = event.message.trimWhiteSpaceAndResets().removeResets()
```
Is there a reason you need the string operations here? I dont see any

j10a1n15 replied:
I just thought "what if the inv check goes wrong" so thats why I added the check from the chat

hannibal002 replied:
Oh no, I meant why you use `trimWhiteSpaceAndResets().removeResets()` there. The string in the regex looks already like the original chat message to me.

j10a1n15 replied:
Ooh, sure imma remove it

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt:
```diff
@@ -0,0 +1,103 @@
+        pattern.matchMatcher(message) {
+            val power = group("power")
+            currentPower = Powers.entries.find { power.contains(it.power) } ?: Powers.UNKNOWN
+            savePower(currentPower!!)
```
Please avoid !! in kotlin. if not directly coming from java code there is no good reason to use it. in this case you can use .also{. Same thing below in the file

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt:
```diff
@@ -0,0 +1,103 @@
+}
+
+object MaxwellAPI {
+    var currentPower: Powers? = null
```
Is there a reason we save the current power here? cant we use the storage directly for read access?

j10a1n15 replied:
We can use MaxwellAPI.currentPower instead of Storage.profilespecific.currentPower,
Thats why I left it like that

hannibal002 replied:
This smoother syntax doesn't require you to keep the classes in the same file still. Look at https://github.com/hannibal002/SkyHanni/blob/ed3fca1da9c642ac6bacc2c2254f36d982bd25b2/src/main/java/at/hannibal2/skyhanni/features/garden/GardenAPI.kt#L51

j10a1n15 replied:
Does that get reloaded once the profile switch happens? Coming from js, after a profile switch you'd still work with the old data

hannibal002 replied:
It's the same access to the same fields, my suggested changes here have nothing to do with a profile switch.
In general, all profile specific accesses are only calculated when needed, not stored directly. This function is not a cache, it's just a better to read redirect call.

The underlying data is always stored to the correct profile the user is in at the time the access happens

j10a1n15 replied:
[Tell me if this fits im not sure](https://github.com/hannibal002/SkyHanni/blob/d5baeb9f74425ac09fd48b6c21b18a32ddf554b8/src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt)

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/MayorElection.kt:
```diff
@@ -20,7 +20,8 @@ class MayorElection {
         var rawMayorData: MayorJson? = null
         var candidates = mapOf<Int, MayorJson.Candidate>()
-        var currentCandidate: MayorJson.Candidate? = null
+        var currentCandidate: MayorJson.Candidate? = null  //todo: should it not be called currentMayor?
```
Im pretty sure it can. feel free to fix in this pr

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/MayorElection.kt:
```diff
@@ -33,9 +34,23 @@ class MayorElection {
 
         if (event.repeatSeconds(3)) {
             check()
```
If we have two functions that get called in one block, having one just named "check" is confusing. What gets checked?

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/PartyAPI.kt:
```diff
@@ -15,6 +15,10 @@ object PartyAPI {
     // TODO USE SH-REPO
```
We shuld use repopatterns for the whole class. not sure if this pr is the correct spot though

j10a1n15 replied:
Please not <3

hannibal002 replied:
then we are on the same page here, nice.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/PartyAPI.kt:
```diff
@@ -15,6 +15,10 @@ object PartyAPI {
     private val youJoinedPartyPattern = "§eYou have joined (?<name>.*)'s §eparty!".toPattern()
     private val othersJoinedPartyPattern = "(?<name>.*) §ejoined the party.".toPattern()
+    // §dParty Finder §f> §bSkirtwearer §ejoined the group! (§3Combat Level 46§e)
+    private val kuudraFinderJoinPattern = "§dParty Finder §f> (?<name>.*?) §ejoined the group! \\(§[a-fA-F0-9]+Combat Level (\\d+)§e\\)".toPattern()
```
I think you can remove the brackets around `\d+`.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/data/PurseAPI.kt:
```diff
@@ -13,9 +13,12 @@ import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
-    private var currentPurse = 0.0
     private var inventoryCloseTime = 0L
 
+    companion object {
```
We can make the whole class an object instead

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboard.kt:
```diff
@@ -0,0 +1,114 @@
+//  - beacon power
+//  - skyblock level
+//  - more bg options (round, blurr, outline)
+//  - countdown events like fishing festival + fiesta when its not on tablist
```
omg yes yes yes

j10a1n15 replied:
Ay stop this will come later on lmao

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboard.kt:
```diff
@@ -0,0 +1,114 @@
+        for (element in config.scoreboardEntries) {
+            lineMap[element.ordinal]?.let {
+                // Hide consecutive empty lines
+                if (config.informationFilteringConfig.hideConsecutiveEmptyLines && it[0].first == "<empty>" && newList.last().first == "") {
```
You can use `.isEmpty()` instead of  `== ""`

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboardUtils.kt:
```diff
@@ -0,0 +1,67 @@
+private val config get() = SkyHanniMod.feature.gui.customScoreboard
+
+object CustomScoreboardUtils {
+    fun mayorNameToColorCode(input: String): String {
```
Feels more consistent in the mayor manager where the api stuff happens. also rename this class then to mayor api

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboardUtils.kt:
```diff
@@ -0,0 +1,67 @@
+    fun getTablistFooter(): String {
+        val tabList = Minecraft.getMinecraft().ingameGUI.tabList as AccessorGuiPlayerTabOverlay
+        if (tabList.footer_skyhanni == null) return ""
+        return tabList.footer_skyhanni.formattedText.replace("§r", "")
```
Maybe use `stripVanillaMessage()` instead of the manual string operation

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboardUtils.kt:
```diff
@@ -0,0 +1,67 @@
+import at.hannibal2.skyhanni.utils.StringUtils.matchMatcher
+import net.minecraft.client.Minecraft
+
+private val config get() = SkyHanniMod.feature.gui.customScoreboard
```
Why is this outside the object scope? feels out of touch

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/Events.kt:
```diff
@@ -0,0 +1,531 @@
+
+private val config get() = SkyHanniMod.feature.gui.customScoreboard
+
+enum class Events(private val displayLine: Supplier<List<String>>, private val showWhen: () -> Boolean) {
```
What events? without reading more into the code i have no idea. more exact names

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/Events.kt:
```diff
@@ -0,0 +1,531 @@
+package at.hannibal2.skyhanni.features.misc.customscoreboard
+
+import at.hannibal2.skyhanni.SkyHanniMod
```
Please dont repeat the hard to read formatting from discord rpc here. move each logic block into its own function and not part of the enum constructors

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/Events.kt:
```diff
@@ -0,0 +1,531 @@
+        {
+            val list = mutableListOf<String>()
+
+            list += ScoreboardData.sidebarLinesFormatted.firstOrNull { it.startsWith("§6Year ") && it.endsWith(" Votes") }
```
You can write a shorter local function to access ScoreboardData.sidebarLinesFormatted in this class

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/CustomScoreboard.kt:
```diff
@@ -0,0 +1,114 @@
+        if (!isCustomScoreboardEnabled()) return
+
+        // Creates the lines
+        display = createLines()
```
How many startsWith() calls does this one function call cause? per tick? this feels super inperformant if im not missing something. Please use a combination of time interval update and scoreboard update, best done with a `dirty` flag that gets checked every 5ish ticks or so

j10a1n15 replied:
[Like this?](https://github.com/hannibal002/SkyHanni/pull/699/commits/97cec4e341daebfb7d6d298fb1053c4993ffe620)

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/InformationGetter.kt:
```diff
@@ -0,0 +1,223 @@
+            heat = "§c♨ 0"
+        }
+
+        // I know this could maybe be solved better but honestly idc anymore
```
Yes this needs a better apprach

j10a1n15 replied:
I am working on it :)

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/InformationGetter.kt:
```diff
@@ -0,0 +1,223 @@
+        }
+
+        // remove trapper mob location
+        unknownLines = unknownLines.filter { sidebarLines.nextAfter("Tracker Mob Location:", 1) != it }
```
For the logic that happens in this line you could write a utils function `removeNextAfter(` to have fewer filter and nextAfter calls, fewer code duplication

j10a1n15 replied:
the original line on which we call nextAfter() will absolutely not exist here anymore
thats why I call sidebarLines.nextAfter so we know the line after here "Tracker Mob Location"

and then we can filter that out

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/RenderBackground.kt:
```diff
@@ -0,0 +1,90 @@
+
+        position.set(
+            Position(
+                when (config.displayConfig.alignment.alignRight) {
```
Is when truly better than if here? im not sure

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/ScoreboardPattern.kt:
```diff
@@ -0,0 +1,9 @@
+
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
+
+object ScoreboardPattern {
```
As long as this class only holds one pattern, it surely can be moved into CustomScoreboardUtils

j10a1n15 replied:
Im currently counting 44 on my end, will push soon:tm:

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/ScoreboardElements.kt:
```diff
@@ -0,0 +1,413 @@
+private val config get() = SkyHanniMod.feature.gui.customScoreboard
+
+// Stats / Numbers
+var purse = "0"
```
Im confused why you have those elements here. stuff like purse and location already exists, everything else should be extracted into their respective api classes.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/customscoreboard/ScoreboardElements.kt:
```diff
@@ -0,0 +1,413 @@
+private val config get() = SkyHanniMod.feature.gui.customScoreboard
+
+// Stats / Numbers
+var purse = "0"
```
Im confused why those values exist here. purse and location exists already as api, the other stuff should get saved in their api classes,  or create ones if they dont exist yet.

j10a1n15 replied:
the issue with purse is, when the purse is over the int limit, and I need to translate it, well

j10a1n15 replied:
also does location have the symbol at the front?

hannibal002 replied:
![image](https://github.com/hannibal002/SkyHanni/assets/24389977/fc0a4222-cae4-4ac1-8669-742746114e01)

idk wdym with location and symbol.

j10a1n15 replied:
the symbol for the location line, like here 
![image](https://github.com/hannibal002/SkyHanni/assets/45315647/a7d7e40a-c938-444d-a9f2-8207251bfa13)

hannibal002 reviewed (changes requested):
I started to review for the first round. IntelliJ stopped showing my requested changes correctly so i guess ill pause for now. 
overall the code looks maintainable and has its correct spot, some classes are way too big still and im not that happy about the big enums yet. not sure what a better solution can look like though. Im confident we can get this big update out in a week or so.

j10a1n15 committed:
Featuretoggle gone

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
Featuretoggle is back (in main config)

j10a1n15 committed:
config.enabled at the end

j10a1n15 committed:
Removed profile switch in favour of bitsapi

j10a1n15 committed:
Typealias very cool!

j10a1n15 committed:
Split up MaxwellAPI into MaxwellPowers and the rest

j10a1n15 committed:
Moved category below global vars

j10a1n15 committed:
removed most config.

j10a1n15 committed:
renamed currentCandidate to currentMayor

j10a1n15 committed:
I think I updated maxwell to your liking

j10a1n15 committed:
Removed when

j10a1n15 committed:
removeNextAfter()

j10a1n15 committed:
MayorElection -> MayorAPI + mayorNameToColorCode()

j10a1n15 committed:
MayorAPI changes in scoreboardelements

j10a1n15 committed:
Now using .isEmpty()

j10a1n15 committed:
Removed old useless method

j10a1n15 committed:
Renamec check -> checkHypixelAPI

j10a1n15 committed:
StripVanillaMessage :+1:

j10a1n15 committed:
PurseAPI is now an object

j10a1n15 committed:
Removed these brackets

j10a1n15 committed:
Config name hideIrrelevantLines cleanup

j10a1n15 committed:
Updating the scoreboard less

j10a1n15 committed:
Renamed Arrow Enum to quiverArrowType

j10a1n15 committed:
Moved var config into scope

j10a1n15 committed:
Welcome in Regex Hell, + more changes

j10a1n15 committed:
events no longer hard to read

j10a1n15 committed:
Some changes (i dont remember)

j10a1n15 committed:
Fixed some things, added boss damage thing

j10a1n15 committed:
Added effigies pattern

j10a1n15 committed:
Added flight duration, fixed bloodeffigies regex

j10a1n15 committed:
fixed using wrong effigy pattern

j10a1n15 committed:
Removed effigies from knownLines list

j10a1n15 committed:
Fixed regexes with lines that starts with a space

j10a1n15 committed:
Fixed bank detection when coop

j10a1n15 committed:
Commented these out because I need them

j10a1n15 committed:
Added server rs pattern

j10a1n15 committed:
Hopefully not breaking anything with this

walkerselby commented on src/main/java/at/hannibal2/skyhanni/config/Features.java:
```diff
@@ -105,7 +105,7 @@ public String getTitle() {
 
     @Expose
-    @Category(name = "Events", desc = "Stuff that is not always available.")
+    @Category(name = "ScoreboardEvents", desc = "Stuff that is not always available.")
```
I don't think this was supposed to change -- this is the actual Events Config,

j10a1n15 replied:
Oh ty, yeah it wasnt supposed to

j10a1n15 committed:
Fixed walkers review thing

walkerselby commented on src/main/java/at/hannibal2/skyhanni/data/MaxwellPowers.kt:
```diff
@@ -0,0 +1,41 @@
+package at.hannibal2.skyhanni.data
+
+enum class MaxwellPowers(val power: String) {
```
Is there a benefit to doing this as an enum rather than something that could be implemented through the Repo, allowing for quicker responses to Hypixel changes?

j10a1n15 replied:
I'd probably be better if it was a repo thing, buttt I dont wanna do it regex is alr killing mee

hannibal002 replied:
maxwell powers needs to be a repo.

j10a1n15 committed:
Fixed one more

j10a1n15 committed:
another one

walkerselby commented on src/main/java/at/hannibal2/skyhanni/data/QuiverAPI.kt:
```diff
@@ -0,0 +1,186 @@
+    private val clearedPattern by group.pattern("cleared", "§aCleared your quiver!")
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
```
could this be adjusted to return, so the following patterns are not checked once one already runs?

j10a1n15 committed:
another onee

j10a1n15 committed:
Less pattern checking in quiverapi

j10a1n15 committed:
Number math fix when getting bits

j10a1n15 committed:
Fixed location pattern, added raffle pattern

j10a1n15 committed:
Added clarifying comments

j10a1n15 committed:
Updated comments

j10a1n15 committed:
zone event now uses pattern

j10a1n15 committed:
And now also filters them with regex

j10a1n15 committed:
Now using PurseAPI's pattern

j10a1n15 committed:
Added GroupNotFoundException

j10a1n15 committed:
fixed heat & compass

j10a1n15 committed:
Added dojo & mithril pattern, fixed compass part 2

j10a1n15 committed:
Made comment better

j10a1n15 committed:
Now using ery's patterns

j10a1n15 committed:
Changed up bank and gems getting

j10a1n15 committed:
Changed filename of InformationGetter

j10a1n15 committed:
Now hiding objective with pattern

j10a1n15 committed:
Redid the pattern filtering

j10a1n15 committed:
Fixed coins detection

j10a1n15 committed:
Fixed bits pattern not working when there is no space between normal amount and earned

j10a1n15 committed:
Added nearby players pattern

j10a1n15 committed:
Magmaboss cleanup

j10a1n15 committed:
Now hiding the redstone line

j10a1n15 committed:
Added trapper

j10a1n15 committed:
Fixed objective + purse pattern

j10a1n15 committed:
Removed UnknownGroupException

j10a1n15 committed:
Fixed Slayer Quest

j10a1n15 committed:
Fixed Slayer and Jacob Pattern

j10a1n15 committed:
Fixed trapper mob next line hiding

j10a1n15 committed:
Fixed Dungeon Lines + other things

j10a1n15 committed:
Removed random (+50) at the end of powder gain

j10a1n15 committed:
Kuudras objective line is plural

j10a1n15 committed:
QuiverAPI sound distance check, dungeon pattern fixes

j10a1n15 committed:
Removed Rounded Rectangle in favour of #851

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
Added spooky pattern

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
Unclaimed bits color less dark now

j10a1n15 committed:
Changed color again

j10a1n15 committed:
There can be two diff wind symbols at once yay

j10a1n15 committed:
Fixed bitsapi pattern

j10a1n15 committed:
Better together is now pink

j10a1n15 committed:
Added goblin raid pattern

j10a1n15 committed:
Take the L

j10a1n15 committed:
Added numberformatting

j10a1n15 committed:
Fixed nearbyplayers pattern when N/A

j10a1n15 committed:
Now hiding the restarting line

j10a1n15 committed:
Fixed bits pattern when decimal place

j10a1n15 committed:
Probably fixed cookie ate pattern

j10a1n15 committed:
Fixed magmaboss lines

j10a1n15 committed:
Fixed empty list in magmaboss

j10a1n15 committed:
Fixed bits pattern again

j10a1n15 committed:
Fixed redstonepattern

j10a1n15 committed:
Fixed time pattern + time updates faster now

j10a1n15 committed:
Bumped arrow shoot distance to 5

j10a1n15 committed:
Fixed lobbycode sometimes showing 0

j10a1n15 committed:
Might need to revert this idk

j10a1n15 committed:
Revert "Might need to revert this idk"

This reverts commit 5b7388d758e292262ce623a9d58b255da92f1ca5.

j10a1n15 committed:
Added redstone lines

j10a1n15 committed:
Removed debug msg when you ate a cookie

j10a1n15 committed:
Another Bits fix

j10a1n15 committed:
Added winter event patterns

j10a1n15 committed:
Added earned amount to purse

j10a1n15 committed:
Fixed purse pattern

j10a1n15 committed:
Added new years event pattern

j10a1n15 committed:
Fixed purse pattern when decimal num

j10a1n15 committed:
Added floor3 guardians

j10a1n15 committed:
Fixed purse with decimal place again

j10a1n15 committed:
Fixed profile change thingy idk

j10a1n15 committed:
Fixed Purse Pattern when you earn coins with decimals

j10a1n15 committed:
Now also show purse where there is no purse line

j10a1n15 committed:
Fixed bits numbers

j10a1n15 committed:
That was weird

j10a1n15 committed:
Made fame enum its own file

j10a1n15 committed:
Fixed visiting Pattern

j10a1n15 committed:
Fixed Purse showing nothing

j10a1n15 committed:
Fixed windcompass pattern again part 327

j10a1n15 committed:
Made elements enum cooler

j10a1n15 committed:
Changed "None 0 Arrows" to "No Arrows selected"

j10a1n15 committed:
Fixed powderpattern

j10a1n15 committed:
Changed some patterns

j10a1n15 committed:
Added bars to start and end of compass

j10a1n15 committed:
Added kuudra patterns (untested)

j10a1n15 committed:
Fixed northstar detection + added short num support

j10a1n15 committed:
Fixed objective line

j10a1n15 committed:
Pattern fixes

j10a1n15 committed:
Updated Purse Pattern

j10a1n15 committed:
Attempt at fixing empty wind line

j10a1n15 committed:
Updated purse pattern key to the same as calm's key

j10a1n15 committed:
Fixed bits pattern

j10a1n15 committed:
Now hiding cookie on bingo

j10a1n15 committed:
Fixed objective line when its named quest

j10a1n15 committed:
Now hiding bits on bingo

j10a1n15 committed:
Hopefully fixed all future purse pattern problems

j10a1n15 committed:
Fixed your goblin kills pattern

j10a1n15 committed:
MaxwellAPI rework thingy

j10a1n15 committed:
Fixed mp number not updating

j10a1n15 committed:
Formatting

j10a1n15 committed:
Formatting

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
MaxwellAPI changes

j10a1n15 committed:
More changes

j10a1n15 committed:
Quiver display changes

j10a1n15 committed:
Fixed lobbycode not showing when the server is closing

j10a1n15 committed:
Maybe fixed multiline objectives

j10a1n15 committed:
Fixed wind compass line

j10a1n15 committed:
Option to customize QUIVERAPI max sound distance

j10a1n15 committed:
Fixed wind pattern part 654

j10a1n15 committed:
Removed uppercase in pattern key

j10a1n15 committed:
More windcompass pattern fixes (i love this game)

j10a1n15 committed:
fixed dojo points pattern

j10a1n15 committed:
Fixed title/footer misalignment when scale thats not 1

j10a1n15 committed:
Fixed spooky pattern bruh

j10a1n15 committed:
Fixed magmabossbar pattern

j10a1n15 committed:
Removed more uppercase keys

j10a1n15 committed:
Fixed time pattern when random char the end

j10a1n15 committed:
Changed some things in infofiltering config

j10a1n15 committed:
Added option to render from cache on island switch

j10a1n15 committed:
Fixed bits to claim going up when buying smth

j10a1n15 committed:
Fixed cleanup and pasting patterns

j10a1n15 committed:
Replaced earned with greedy. why? because fuck hypixel thats why

j10a1n15 committed:
Fixed some garden lines + flight duration

j10a1n15 committed:
Formatting change

j10a1n15 committed:
"useless trailing comma"

j10a1n15 committed:
Bingoth

j10a1n15 committed:
Island diplays name now using displayname

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
Added darkauction event

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
No longer updating mp in dungeons

j10a1n15 committed:
Fixed empty lists in events

j10a1n15 committed:
Merge branch 'hannibal002:beta' into customscoreboard

j10a1n15 committed:
shortened the isenabled fun

j10a1n15 committed:
Merge remote-tracking branch 'origin/customscoreboard' into customscoreboard

j10a1n15 committed:
Removed uppercase in pattern key names

hannibal002 committed:
Merge branch 'beta' into customscoreboard

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/SkyHanniMod.kt

j10a1n15 committed:
Added byNameOrNull to maxwellapi
